### ISSUE TITLE
Problem with: cross-compiling for raspberry pi

### ISSUE THREAD

**Issue by eLEcTRiCZiTy** [CGAL/cgal-swig-bindings#55]
I have a problem with cross-compiling. Any ideas on a solution?

crosspi OS: Ubuntu 14.10
cross-compiler: https://github.com/raspberrypi/tools/tree/master/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian
cross-compiler setup: http://airwiki.ws.dei.polimi.it/index.php/Cross-compiling_for_the_RaspberryPi

crosscompiling user "crosspi", crosscompiling dir "rpi", raspberry pi root in "rootfs" (created with rsync)

___
git command: git clone https://code.google.com/p/cgal-bindings

CMAKE command:
```
cmake -D CGAL_DIR=$HOME/rpi/rootfs/usr/lib/CGAL -D CMAKE_TOOLCHAIN_FILE=$HOME/rpi/toolchain-rpi.cmake .
```
cmake configure and generate... but print warnings likes:
```
CMake Warning at /usr/share/cmake-2.8/Modules/UseSWIG.cmake:210 (add_library):
  Cannot generate a safe runtime search path for target
  _CGAL_Box_intersection_d because files in some directories may conflict
  with libraries in implicit directories:
    runtime library [libgmpxx.so] in /usr/lib/arm-linux-gnueabihf may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib/arm-linux-gnueabihf
    runtime library [libmpfr.so] in /usr/lib/arm-linux-gnueabihf may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib/arm-linux-gnueabihf
    runtime library [libgmp.so] in /usr/lib/arm-linux-gnueabihf may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib/arm-linux-gnueabihf
    runtime library [libboost_thread-mt.so] in /usr/lib may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib
    runtime library [libGLU.so] in /usr/lib/arm-linux-gnueabihf may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib/arm-linux-gnueabihf
    runtime library [libGL.so] in /usr/lib/arm-linux-gnueabihf may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib/arm-linux-gnueabihf
    runtime library [libX11.so] in /usr/lib/arm-linux-gnueabihf may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib/arm-linux-gnueabihf
    runtime library [libXext.so] in /usr/lib/arm-linux-gnueabihf may be hidden by files in:
      /home/crosspi/rpi/rootfs/usr/lib/arm-linux-gnueabihf
  Some of these libraries may not be found correctly.
Call Stack (most recent call first):
  cmake/Modules/SWIG_CGAL_Macros.cmake:83 (SWIG_ADD_MODULE)
  SWIG_CGAL/Box_intersection_d/CMakeLists.txt:5 (ADD_SWIG_CGAL_PYTHON_MODULE)
```
___
MAKE:
```
crosspi@ubuntu:~/rpi/cgal-bindings$ make
Scanning dependencies of target CGAL_Kernel_cpp
[  1%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Point_2.cpp.o
[  3%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Weighted_point_2.cpp.o
[  4%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Segment_2.cpp.o
[  6%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Triangle_2.cpp.o
[  8%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Ray_2.cpp.o
[  9%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Direction_2.cpp.o
[ 11%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Line_2.cpp.o
[ 12%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Vector_2.cpp.o
[ 14%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Polygon_2.cpp.o
[ 16%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Bbox_2.cpp.o
[ 17%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Point_3.cpp.o
[ 19%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Weighted_point_3.cpp.o
[ 20%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Sphere_3.cpp.o
[ 22%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Plane_3.cpp.o
[ 24%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Segment_3.cpp.o
[ 25%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Line_3.cpp.o
[ 27%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Triangle_3.cpp.o
[ 29%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Tetrahedron_3.cpp.o
[ 30%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Direction_3.cpp.o
[ 32%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Ray_3.cpp.o
[ 33%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Vector_3.cpp.o
[ 35%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Bbox_3.cpp.o
[ 37%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Object.cpp.o
[ 38%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/global_functions.cpp.o
/tmp/ccK8ERLv.s: Assembler messages:
/tmp/ccK8ERLv.s:10635: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
/tmp/ccK8ERLv.s:10674: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
/tmp/ccK8ERLv.s:10836: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
/tmp/ccK8ERLv.s:10877: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
/tmp/ccK8ERLv.s:182091: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
/tmp/ccK8ERLv.s:182127: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
/tmp/ccK8ERLv.s:256924: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
/tmp/ccK8ERLv.s:256960: Warning: swp{b} use is deprecated for ARMv6 and ARMv7
[ 40%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Origin.cpp.o
[ 41%] Building CXX object SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/Iso_rectangle_2.cpp.o
make[2]: *** No rule to make target '/usr/lib/arm-linux-gnueabihf/libgmpxx.so', needed by 'lib/libCGAL_Kernel_cpp.so'.  Stop.
CMakeFiles/Makefile2:77: recipe for target 'SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/all' failed
make[1]: *** [SWIG_CGAL/Kernel/CMakeFiles/CGAL_Kernel_cpp.dir/all] Error 2
Makefile:82: recipe for target 'all' failed
make: *** [all] Error 2
crosspi@ubuntu:~/rpi/cgal-bindings$
```
(if I symlink missing lib from rpi to crosspi /usr/lib/ then it works... maybe... problem is finding libs whem make compiling.. it finding libs in host system and fail ..because libs are in ~/rpi/rootfs/.. cmake crosscompile file is configured for finding libs in rpi)

>sorry for my bad english

**Comment by lrineau** [comment]
Note that, so far, the CGAL project do not support ARM platforms, neither cross-compilers. But we are pleased to try and help anyway.

Thanks for the detailed report. That is useful. The useful error message is the following:
```
make[2]: *** No rule to make target '/usr/lib/arm-linux-gnueabihf/libgmpxx.so', needed by 'lib/libCGAL_Kernel_cpp.so'. Stop.
```
For some reason, your toolchain file is referring to a file `/usr/lib/arm-linux-gnueabihf/libgmpxx.so` that does not exist. If you cannot find a way to get it, you can set the CMake option `WITH_GMPXX` to `OFF` or `FALSE` to deactivate the CGAL support for GMPXX.

**Comment by eLEcTRiCZiTy** [comment]
My mistake... crosscompiling toolchain libraries must contain the same libraries as Raspberry Pi and Cmake. Cmake crosscompile file contain path to libs and includes. Make (crosscompiler-gcc,...) does not know where these libraries are. Custom crosscompile toolchain with symlinks to Raspbian libraries will work. I used https://github.com/raspberrypi/tools crosscompiler and copy libraries (/lib and /usr/lib) from Raspberry Pi into <crosscompiler>/arm-linux-gnueabihf/libc. It's a bit ugly and impractical solution but it works. Crosscompiling lasted about an hour and in the worst case need about 5 gigabytes of memory (module Mesh_3). 64-bit version of the crosscompiler (and Linux) is a must. I used virtualized Ubuntu 14.10 64bit, virtualized single core cpu 2.8GHz (one thread,one core from Intel Core i7), 2GB RAM and 10GB swap.
